compiler/rustc_middle/src/ty/fast_reject.rs

```diff
@@ -269,6 +272,9 @@ impl DeepRejectCtxt {
                 }
                 _ => false,
             },
+            ty::Pat(obl_ty, _) => {
+                matches!(k, &ty::Pat(impl_ty, _) if self.types_may_unify(obl_ty, impl_ty))
```

Comment:
add a FIXME to make this matching more exact -- we can definitely do deep matching on the pattern values too